### PR TITLE
Add aria-describedby for question description and errors for input fields

### PR DESCRIPTION
**Pull request by jeanniefu**
### Description
Adds question level `aria-describedby` attributes (e.g. question description and question-level errors) to each input of that question.

Previously, the screen reader would only read out field-level errors but not question-level errors (See https://github.com/seattle-uat/civiform/issues/1878#issuecomment-1136572406).

Note that this PR does not address radio buttons, checkboxes, and misses some corner cases for enumerators. I will follow up with separate fixes.

### Checklist
- [x] Created tests which fail without the change (if possible)
- [ ] Extended the README / documentation, if necessary

### Issue(s)
#1878, #1879

### CODE REVIEW

File: server/app/views/components/FieldWithLabel.java [hunk 1]
```diff
@@ -221,6 +227,18 @@ public FieldWithLabel setScreenReaderText(String screenReaderText) {
...
   }
 
+  /** Set the list of HTML tag IDs that should be used for a11y descriptions. */
+  public FieldWithLabel setAriaDescribedByIds(List<String> ariaDescribedByIds) {
+    this.ariaDescribedByIds = new ArrayList<>(ariaDescribedByIds);
+    return this;
+  }
+
+  /** Set whether the question this field belongs to has errors. */
+  public FieldWithLabel setHasQuestionErrors(boolean hasQuestionErrors) {
```
Comment (clouser2) [inline server/app/views/components/FieldWithLabel.java:237]: FieldWithLabel is used for rendering fields outside of the context of an application (e.g. within the Admin UI when defining question / program types). Putting questions-specific logic feels like it might be introducing a bit too much coupling with the applicant experience.

What about something like `forceIsInvalid()` to ensure that `aria-invalid` is set even if there aren't any field-specific validation errors?

File: server/app/views/components/FieldWithLabel.java [hunk 2]
```diff
@@ -283,10 +301,16 @@ public ContainerTag getContainer() {
...
+    fieldTag.condAttr(hasFieldErrors || hasQuestionErrors, "aria-invalid", "true");
+
+    var allAriaDescribedByIds = new ArrayList<String>(ariaDescribedByIds);
     if (hasFieldErrors) {
-      fieldTag.attr("aria-invalid", "true");
-      fieldTag.attr("aria-describedBy", fieldErrorsId);
+      allAriaDescribedByIds.add(0, fieldErrorsId);
     }
+    fieldTag.condAttr(
+        !allAriaDescribedByIds.isEmpty(),
```
Comment (clouser2) [inline server/app/views/components/FieldWithLabel.java:311]: Is a blank value acceptable for `aria-describedBy`? If so, we could unconditionally set this.

**Review by clouser2 — CHANGES_REQUESTED** [review]
Love the direction overall! Had a few questions around whether when we want to associate top-level question errors with specific fields.

File: server/app/views/questiontypes/ApplicantQuestionRendererImpl.java [hunk 3]
```diff
@@ -71,27 +89,36 @@ public final Tag render(ApplicantQuestionRendererParams params) {
...
+      ariaDescribedByIds.add(0, requiredErrorId);
     }
 
     return div()
-        .withId(question.getContextualizedPath().toString())
+        .withId(id)
         .withClasses(Styles.MX_AUTO, Styles.MB_8, getReferenceClass(), getRequiredClass())
         .with(questionTextDiv)
-        .with(renderTag(params, validationErrors));
+        .with(renderTag(params, validationErrors, ariaDescribedByIds, hasQuestionErrors));
```
Comment (bion) [inline server/app/views/questiontypes/ApplicantQuestionRendererImpl.java:122]: suggestion (blocking): it's generally a good practice to avoid passing mutable collections around when avoidable. in this class you could make `ariaDescribedByIds` an `ImmutableList.Builder<String>` and change this line to
```suggestion
        .with(renderTag(params, validationErrors, ariaDescribedByIds.build(), hasQuestionErrors));
```

File: server/app/views/questiontypes/EnumeratorQuestionRenderer.java [hunk 4]
```diff
@@ -83,6 +90,11 @@ protected Tag renderTag(
                             messages.at(
                                 MessageKey.ENUMERATOR_BUTTON_ADD_ENTITY.getKeyName(),
                                 localizedEntityType)))
+                    .condAttr(hasQuestionErrors, "aria-invalid", "true")
+                    .condAttr(
+                        !ariaDescribedByIds.isEmpty(),
+                        "aria-describedby",
```
[the remark this answers is not in the document]
Comment (bion) [reply server/app/views/questiontypes/EnumeratorQuestionRenderer.java:96]: doesn't look like it https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-describedby

**Review by bion — CHANGES_REQUESTED** [review]

File: server/app/views/questiontypes/ApplicantQuestionRendererImpl.java [hunk 5]
```diff
@@ -71,27 +89,36 @@ public final Tag render(ApplicantQuestionRendererParams params) {
             String.format("Unhandled error display mode: %s", params.errorDisplayMode()));
     }
 
+    boolean hasQuestionErrors = false;
     ImmutableSet<ValidationErrorMessage> questionErrors =
         validationErrors.getOrDefault(question.getContextualizedPath(), ImmutableSet.of());
     if (!questionErrors.isEmpty()) {
       // Question error text
+      hasQuestionErrors = true;
```
[the remark this answers is not in the document]
Comment (jeanniefu) [reply server/app/views/questiontypes/ApplicantQuestionRendererImpl.java:97]: Good points! I checked with the a11y team, and they said we should NOT associate the question level errors and descriptions with the sub fields. 

They also recommend we group the composite fields (like address, name) using fieldset and legend, and associate the question level errors/descriptions with the fieldset. Related issue: https://github.com/seattle-uat/civiform/issues/2618. Also, we may want to change the way we handle required fields and errors based on a11y team's feedback (https://github.com/seattle-uat/civiform/issues/2629), which might also change my approach here.

I think I should change this PR to just affect the non-composite fields. Will comment again here once it's ready.

**Comment by jeanniefu** [comment]
This was redone and submitted as #3354
